shotover/src/codec/cassandra.rs

```diff
@@ -673,65 +705,128 @@ impl CassandraEncoder {
-                        dst[header_start..header_start + 3].copy_from_slice(payload_len);
-                        dst[header_start + 3..header_start + 6]
-                            .copy_from_slice(&crc24(payload_len).to_le_bytes()[..3]);
+                            // add header length & header crc
+                            let payload_len = &payload_len.to_le_bytes()[..3];
+                            dst[header_start..header_start + 3].copy_from_slice(payload_len);
+                            dst[header_start + 3..header_start + 6]
+                                .copy_from_slice(&crc24(payload_len).to_le_bytes()[..3]);
 
-                        // add payload crc
-                        dst.extend_from_slice(&crc32(&dst[payload_start..]).to_le_bytes());
+                            // add payload crc
+                            dst.extend_from_slice(&crc32(&dst[payload_start..]).to_le_bytes());
+                        }
                     }
                     Compression::Lz4 => {
-                        let header_start = dst.len();
+                        let mut envelope_bytes = self.encode_envelope(m, Compression::None)?;
 
-                        dst.extend_from_slice(&[0, 0, 0, 0, 0, 0, 0, 0]);
-                        let payload_start = dst.len();
+                        if get_maximum_output_size(envelope_bytes.len()) > PAYLOAD_SIZE_LIMIT {
+                            while !envelope_bytes.is_empty() {
+                                let header_start = dst.len();
 
-                        // TODO we should not be encoding small frames
-                        let (uncompressed_len, compressed_len) =
-                            self.encode_compressed_payload(dst, m, payload_start)?;
+                                dst.extend_from_slice(&[0, 0, 0, 0, 0, 0, 0, 0]);
+                                let payload_start = dst.len();
 
-                        if compressed_len > PAYLOAD_SIZE_LIMIT {
-                            todo!("non self contained frames not yet implemented.")
-                        }
+                                let payload_bytes = envelope_bytes
+                                    .split_to(envelope_bytes.len().min(PAYLOAD_SIZE_LIMIT - 1));
 
-                        if uncompressed_len > PAYLOAD_SIZE_LIMIT {
-                            todo!("non self contained frames not yet implemented.")
-                        }
+                                let (mut uncompressed_len, mut compressed_len) = self
+                                    .encode_compressed_payload_into_buffer(
+                                        dst,
+                                        &payload_bytes,
+                                        payload_start,
+                                    )?;
 
-                        let mut header =
-                            (compressed_len) as u64 | ((uncompressed_len as u64) << 17);
+                                if compressed_len >= PAYLOAD_SIZE_LIMIT {
```

Comment:
I dont understand the logic in this branch, can you explain what its for?

It seems like: If the compressed payload needs to be split up fall back to an uncompressed payload.
But I dont understand why we would want to do that.

Reply:
https://github.com/shotover/shotover-proxy/pull/1143

This is checking the same thing as this PR but I will just check against the uncompressed length to make it more clear.